### DISCUSSION TITLE
Help With FUSE Overlay-fs on Gentoo (OpenRC)

### DISCUSSION THREAD

**Discussion by ThisNekoGuy** [filips123/PWAsForFirefox#472] [Q&A]
- https://gist.github.com/filips123/29fb511a01ee8016a927a614f32979d3#non-systemd-os

The guide for doing this on non-systemd systems doesn't explain two things:

1. `replacing %RUNTIME-PATH% with the path to your normal Firefox runtime`
1b. This doesn't explain where the "normal FireFox runtime" is located (I use LibreWolf)
2. `When logging out, your script should:`
2b. Doesn't explain **_how_** to unmount on **_logout_**, per user $HOME path being logged out (no absolute paths)
2c. Doesn't explain how to destroy the `upper/` and `work/` directories on log out

Can I get some clarification on these issues? (Also assuming the script is supposed to be in `/etc/profile.d/`)

**Comment by filips123** [comment]
> This doesn't explain where the "normal FireFox runtime" is located (I use LibreWolf)

It is where your Firefox/LibreWolf is installed. So, for LibreWolf, that is probably `/usr/lib/librewolf` or `/usr/lib64/librewolf`.

> Doesn't explain how to unmount on logout

It depends on your init system. For systemd, it's just `ExecStop`. For OpenRC, there is probably also a way to run something when the service stops.

> Doesn't explain how to destroy the `upper/` and `work/` directories on log out

You can just remove them.

Also, I'm not sure if this will work, but you can try to completely skip that logging out part. Maybe the OS will automatically unmount the directories on logout, but I'm not sure.

**Reply by ThisNekoGuy** [reply]
> For OpenRC, there is probably also a way to run something when the service stops.

OpenRC doesn't support user-specific services like systemd does, and I'm not sure if they would only get called on _shutdown_ rather than logout and would only hypothetically affect hard-coded directories (so no `$HOME` variables would work as intended there)

> You can just remove them.

I moreso meant in an automated way, but I sort of mentioned this with the service issue already

**Reply by filips123** [reply]
Well, then you can try if just skipping the logout part still works. If it doesn't, maybe you can still do something with services on shutdown, but unfortunately I'm not familiar with OpenRC.

**Reply by ThisNekoGuy** [reply]
So, I tried setting up fuse-overlayfs and the extension refuses to accept it as a valid runtime for some reason; it remains like this:
<details><summary>Screenshot:</summary>

![Screenshot_20240306_212014](https://github.com/filips123/PWAsForFirefox/assets/28468662/67d6ff53-2a55-43bd-8ae3-4312806e9b30)
</details>

**Reply by filips123** [reply]
What's the content of `$HOME/.local/share/firefoxpwa/runtime`? Does it contain Firefox? Specifically, does it contain the `firefox` binary and `application.ini`?

**Reply by ThisNekoGuy** [reply]
It contains the `librewolf` binary and `application.ini`

**Reply by filips123** [reply]
You also need to symlink `librewolf` with `firefox` binaries.

**Reply by ThisNekoGuy** [reply]
After doing that I finally got past the installation screen but now, when I install and launch a PWA, it doesn't actually load the site :skull_and_crossbones: 
I just get this (the search engine starting page):

<details><summary>Screenshot:</summary>

![Screenshot_20240307_144134](https://github.com/filips123/PWAsForFirefox/assets/28468662/dcaa5e5e-8225-485a-ac0c-5ad99553e20b)
</details>

**Reply by filips123** [reply]
Have you enabled "Always patch runtime and profile" in the extension settings? If yes, please also check [the runtime logs](https://pwasforfirefox.filips.si/help/troubleshooting/#obtaining-runtime-logs).

**Reply by ThisNekoGuy** [reply]
I do.

This was the only thing I was able to find in there that seemed to indicate anything I'd imagine to be related?
<details><summary>Screenshot:</summary>

![Screenshot_20240307_155811](https://github.com/filips123/PWAsForFirefox/assets/28468662/0159fdbb-a2fe-4f85-a32d-17a5ee667234)
</details>

**Reply by filips123** [reply]
Yeah, this means that the PWAsForFirefox wasn't loaded, but I don't know why... Please check if files `~/.local/share/firefoxpwa/runtime/_autoconfig.cfg` and `~/.local/share/firefoxpwa/runtime/defaults/pref/autoconfig.js` exist. Also, try to run "Patch profiles and runtime" from the extension settings, and then make sure you close any existing PWA windows before reopening them.

**Reply by ThisNekoGuy** [reply]
`Patch profiles and runtime` was already set
The `~/.local/share/firefoxpwa/runtime/_autoconfig.cfg` and `~/.local/share/firefoxpwa/runtime/defaults/pref/autoconfig.js` files don't exist :/

**Reply by filips123** [reply]
That's interesting... It seems that runtime wasn't patched, but I still don't know why.

What happens if you try to patch the runtime with `firefoxpwa runtime patch`? If that doesn't work, please also try to manually copy that two files from `/usr/share/firefoxpwa/userchrome/runtime` to (the correct subdirectory of) `~/.local/share/firefoxpwa/runtime`.

Also, do upper and work directories under `$XDG_RUNTIME_DIR/firefoxpwa/overlayfs` (or where you've set them) contain anything?

**Reply by ThisNekoGuy** [reply]
`firefoxpwa runtime patch` seems to copy the files from `/usr/share/firefoxpwa/userchrome/runtime`.

Terminal output:
`firefoxpwa runtime patch`
```
22:13:49 [INFO] Patching the runtime
22:13:49 [INFO] Runtime patched!
```

`ls $XDG_RUNTIME_DIR/firefoxpwa/overlayfs`
```
upper/  work/
```
`ls $XDG_RUNTIME_DIR/firefoxpwa/overlayfs/upper/`
```
_autoconfig.cfg  defaults/
```
`ls $XDG_RUNTIME_DIR/firefoxpwa/overlayfs/work`
```
work/
```
`ls $XDG_RUNTIME_DIR/firefoxpwa/overlayfs/work/work/` -> (Blank output)

**Reply by filips123** [reply]
So, it seems that `fuse-overlayfs` doesn't work for some reason. I think that `fuse-overlayfs` requires some additional system dependencies, do you have them installed? Please also check if it works if you try to use it for other things than PWAsForFirefox, maybe try some examples from [here](https://github.com/containers/fuse-overlayfs) or [here](https://www.baeldung.com/linux/overlayfs-usage).

**Reply by ThisNekoGuy** [reply]
Gentoo compiles `fuse-overlayfs` from source, and the dependencies are still on my system, so I'd think I have the proper dependencies already?

And for reference, this is what my login script looks like:
<details><summary>login.sh</summary>

```bash
#!/bin/bash

if [ "$(id -u)" -gt 0 ] && [ -f /usr/bin/fuse-overlayfs ] && [ -f /usr/lib64/librewolf/native-messaging-hosts/firefoxpwa.json ]; then

    if [ ! -d "$HOME/.local/share/firefoxpwa/runtime" ]; then
        mkdir -p "$HOME/.local/share/firefoxpwa/runtime"
    fi

    mkdir -p "$XDG_RUNTIME_DIR/firefoxpwa/overlayfs/upper"; mkdir -p "$XDG_RUNTIME_DIR/firefoxpwa/overlayfs/work"

    fuse-overlayfs -f -o squash_to_uid=$(id -u),squash_to_gid=$(id -g),lowerdir=/usr/lib64/librewolf,upperdir=$XDG_RUNTIME_DIR/firefoxpwa/overlayfs/upper,workdir=$XDG_RUNTIME_DIR/firefoxpwa/overlayfs/work "$HOME/.local/share/firefoxpwa/runtime"
fi
```
</details>

**Reply by filips123** [reply]
Unfortunetely, I don't know why this happens, and I don't have time currently to test it in Gentoo VM. One more thing you can try is to remove `squash_to_uid` and `squash_to_gid`, maybe change directories, and try to simplify that command. Other than that, I don't know what else you could try...

**Reply by ThisNekoGuy** [reply]
If I remove the squash lines, attempting to run PWAs returns permission denied errors (`~/.local/share/firefoxpwa/runtime/` becomes root-owned):
```
22:33:20 [INFO] Patching the runtime
22:33:20 [ERROR] Failed to patch the runtime

Caused by:
    Permission denied (os error 13)
```
"Simplifying" the command to:
```
fuse-overlayfs -f -o squash_to_uid=$(id -u) -o squash_to_gid=$(id -g) -o lowerdir=/usr/lib64/librewolf -o upperdir=$XDG_RUNTIME_DIR/firefoxpwa/overlayfs/upper -o workdir=$XDG_RUNTIME_DIR/firefoxpwa/overlayfs/ "$HOME/.local/share/firefoxpwa/runtime"
```
(Also changed `workdir` to `$XDG_RUNTIME_DIR/firefoxpwa/overlayfs/` because of the double `work` directory we saw earlier)
But this didn't change its behavior at all.

I wouldn't understand why changing the `upper` and `work` locations would do anything though, because those are both user-owned... I'm really puzzled as to how this is happening.

With squash included, running the PWA command directly via terminal returns this:
```
user@GENTOO ~> /usr/bin/firefoxpwa site launch 01HRDA5VEXYCSCT40EFDS469B3 --protocol                                                                                                                                
22:46:00 [INFO] Patching the runtime
22:46:00 [INFO] Runtime patched!
22:46:00 [INFO] Patching the profile
22:46:00 [INFO] Profile patched!
22:46:00 [INFO] Launching the web app
```
To which the browser launches again but not as intended, like before. I'm not exactly sure what about this process isn't working correctly and being able to just use my system browser is ideal. :/
